Beans from transitive dependencies of a direct optional dependency are not injectable

**Describe the bug**
If a direct dependency of a Maven module is marked as `optional`, beans in transitive dependencies of this optional dependency are not injectable.

E.g.:
- module `core` defines `quarkus-liquibase` (+ `h2` and some other basic stuff)
- module `dist` depends on `core`, but with `<optional>true</optional>`
- module `dist` is injecting `LiquibaseFactory` in some bean (e.g. a resource class)
- module `dist` is also injecting a bean from `core`

In this scenario, the bean from `core` is injectable, but `LiquibaseFactory` is not.

**Expected behavior**
Transitive deps of a direct optional dep are still injectable.
Only _transitive_ deps that are declared as optional should not be available/injectable (e.g. if `core` would mark `quarkus-liquibase` as optional)

**Actual behavior**
ArC fails:
```
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type io.quarkus.liquibase.LiquibaseFactory and qualifiers [@Default]
        - java member: com.github.famod.modmono_quarkus.dist.HelloResource#liquibaseFactory
        - declared on CLASS bean [types=[com.github.famod.modmono_quarkus.dist.HelloResource, java.lang.Object], qualifiers=[@Default, @Any], target=com.github.famod.modmono_quarkus.dist.HelloResource]
```

**To Reproduce**
Steps to reproduce the behavior:
1. git clone https://github.com/famod/modmono-quarkus
2. git switch optional-problem
3. mvn clean install (should fail in test with the mentioned exception)
5. Note that there are _no_ classloading problems! Maven adds `core` and all its transitive deps to the classpath (which is correct as per spec).
6. Remove `@Inject` from `HelloResource.liquibaseFactory` -> Test or DevMode will not fail anymore
7. Put back `@Inject` but remove `<optional>` from `dist/pom.xml` -> Test or DevMode will not fail anymore

**Configuration**
```properties
quarkus.http.port=18080

quarkus.datasource.db-kind=h2
quarkus.datasource.username=sa
quarkus.datasource.password=sa
quarkus.datasource.jdbc.url=jdbc:h2:mem:test;DB_CLOSE_DELAY=-1
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: `MINGW64_NT-10.0-18363 XXX 3.0.7-338.x86_64 2019-11-21 23:07 UTC x86_64 Msys`
 - Output of `java -version`: `OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.7+10, mixed mode)`
 - GraalVM version (if different from Java): n/a
 - Quarkus version or git rev: 1.6.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Apache Maven 3.6.3

**Additional context**
```
$ mvn org.apache.maven.plugins:maven-dependency-plugin:3.1.2:tree -f dist/
[INFO] Scanning for projects...
[INFO]
[INFO] -------< com.github.famod.modmono-quarkus:modmono-quarkus-dist >--------
[INFO] Building modmono-quarkus-dist 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-dependency-plugin:3.1.2:tree (default-cli) @ modmono-quarkus-dist ---
[INFO] com.github.famod.modmono-quarkus:modmono-quarkus-dist:jar:1.0-SNAPSHOT
[INFO] +- com.github.famod.modmono-quarkus:modmono-quarkus-core:jar:1.0-SNAPSHOT:compile (optional)
[INFO] |  +- jakarta.enterprise:jakarta.enterprise.cdi-api:jar:2.0.2:compile
[INFO] |  |  +- jakarta.el:jakarta.el-api:jar:3.0.3:compile
[INFO] |  |  +- jakarta.interceptor:jakarta.interceptor-api:jar:1.2.5:compile
[INFO] |  |  \- jakarta.inject:jakarta.inject-api:jar:1.0:compile
[INFO] |  +- io.quarkus:quarkus-jdbc-h2:jar:1.6.1.Final:compile (optional)
[INFO] |  |  +- com.h2database:h2:jar:1.4.197:compile (optional)
[INFO] |  |  \- org.locationtech.jts:jts-core:jar:1.15.0:compile (optional)
[INFO] |  \- io.quarkus:quarkus-liquibase:jar:1.6.1.Final:compile (optional)
[INFO] |     +- org.liquibase:liquibase-core:jar:3.10.0:compile (optional)
[INFO] |     +- org.yaml:snakeyaml:jar:1.26:compile (optional)
[INFO] |     +- io.quarkus:quarkus-jaxb:jar:1.6.1.Final:compile (optional)
[INFO] |     |  +- org.glassfish.jaxb:jaxb-runtime:jar:2.3.3-b02:compile (optional)
[INFO] |     |  |  +- org.glassfish.jaxb:txw2:jar:2.3.3-b02:compile (optional)
[INFO] |     |  |  \- com.sun.istack:istack-commons-runtime:jar:3.0.10:compile (optional)
[INFO] |     |  \- org.jboss.spec.javax.xml.bind:jboss-jaxb-api_2.3_spec:jar:2.0.0.Final:compile
[INFO] |     +- io.quarkus:quarkus-agroal:jar:1.6.1.Final:compile (optional)
[INFO] |     |  +- io.quarkus:quarkus-datasource:jar:1.6.1.Final:compile (optional)
[INFO] |     |  |  \- io.quarkus:quarkus-datasource-common:jar:1.6.1.Final:compile (optional)
[INFO] |     |  +- org.jboss.narayana.jta:narayana-jta:jar:5.10.5.Final:compile (optional)
[INFO] |     |  |  +- org.jboss:jboss-transaction-spi:jar:7.6.0.Final:compile (optional)
[INFO] |     |  |  \- org.jboss.spec.javax.resource:jboss-connector-api_1.7_spec:jar:1.0.0.Final:compile (optional)
[INFO] |     |  +- org.jboss.narayana.jts:narayana-jts-integration:jar:5.10.5.Final:compile (optional)
[INFO] |     |  +- jakarta.transaction:jakarta.transaction-api:jar:1.3.3:compile
[INFO] |     |  +- io.agroal:agroal-api:jar:1.8:compile (optional)
[INFO] |     |  +- io.agroal:agroal-narayana:jar:1.8:compile (optional)
[INFO] |     |  +- io.agroal:agroal-pool:jar:1.8:compile (optional)
[INFO] |     |  \- io.quarkus:quarkus-credentials:jar:1.6.1.Final:compile (optional)
[INFO] |     +- io.quarkus:quarkus-narayana-jta:jar:1.6.1.Final:compile (optional)
[INFO] |     |  +- io.quarkus:quarkus-mutiny:jar:1.6.1.Final:compile (optional)
[INFO] |     |  |  +- io.quarkus:quarkus-smallrye-context-propagation:jar:1.6.1.Final:compile (optional)
[INFO] |     |  |  |  \- io.smallrye:smallrye-context-propagation:jar:1.0.13:compile (optional)
[INFO] |     |  |  |     \- io.smallrye:smallrye-context-propagation-api:jar:1.0.13:compile (optional)
[INFO] |     |  |  \- io.smallrye.reactive:mutiny-context-propagation:jar:0.5.4:compile (optional)
[INFO] |     |  +- io.smallrye:smallrye-context-propagation-jta:jar:1.0.13:compile (optional)
[INFO] |     |  +- io.smallrye.reactive:smallrye-reactive-converter-api:jar:1.0.13:compile (optional)
[INFO] |     |  |  +- org.eclipse.microprofile.reactive-streams-operators:microprofile-reactive-streams-operators-api:jar:1.0.1:compile (optional)
[INFO] |     |  |  \- org.eclipse.microprofile.reactive-streams-operators:microprofile-reactive-streams-operators-core:jar:1.0.1:compile (optional)
[INFO] |     |  \- io.smallrye.reactive:smallrye-reactive-converter-mutiny:jar:1.0.13:compile (optional)
[INFO] |     \- org.osgi:org.osgi.core:jar:6.0.0:compile (optional)
[INFO] +- io.quarkus:quarkus-resteasy:jar:1.6.1.Final:compile
[...]
```
Notice the `(optional)`, but all deps are on the classpath.

PS: The reproducer is rather silly. The real world setup in which we hit this problem makes more sense.

/cc @mickroll

--- a/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/ExecutableOutputOutcomeTestBase.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/ExecutableOutputOutcomeTestBase.java
@@ -22,6 +22,7 @@ import io.quarkus.bootstrap.app.AugmentAction;
 import io.quarkus.bootstrap.app.AugmentResult;
 import io.quarkus.bootstrap.app.CuratedApplication;
 import io.quarkus.bootstrap.app.QuarkusBootstrap;
+import io.quarkus.bootstrap.model.AppModel;
 import io.quarkus.bootstrap.resolver.TsArtifact;
 import io.quarkus.bootstrap.resolver.update.CreatorOutcomeTestBase;
 
@@ -36,11 +37,15 @@ public abstract class ExecutableOutputOutcomeTestBase extends CreatorOutcomeTest
         expectedLib.add(entry.getGroupId() + '.' + entry.getArtifactId() + '-' + entry.getVersion() + '.' + entry.getType());
     }
 
+    protected void assertAppModel(AppModel appModel) throws Exception {
+    }
+
     @Override
     protected void testCreator(QuarkusBootstrap creator) throws Exception {
         System.setProperty("quarkus.package.type", "legacy");
         try {
             CuratedApplication curated = creator.bootstrap();
+            assertAppModel(curated.getAppModel());
             AugmentAction action = curated.createAugmentor();
             AugmentResult outcome = action.createProductionApplication();
 

--- a/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/OptionalDepsTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/OptionalDepsTest.java
@@ -1,0 +1,77 @@
+package io.quarkus.deployment.runnerjar;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import io.quarkus.bootstrap.model.AppArtifact;
+import io.quarkus.bootstrap.model.AppDependency;
+import io.quarkus.bootstrap.model.AppModel;
+import io.quarkus.bootstrap.resolver.TsArtifact;
+import io.quarkus.bootstrap.resolver.TsDependency;
+import io.quarkus.bootstrap.resolver.TsQuarkusExt;
+
+public class OptionalDepsTest extends ExecutableOutputOutcomeTestBase {
+
+    @Override
+    protected TsArtifact modelApp() {
+
+        final TsArtifact extADep = TsArtifact.jar("ext-a-dep");
+        addToExpectedLib(extADep);
+
+        final TsArtifact extAOptionalDep = TsArtifact.jar("ext-a-optional-dep");
+        final TsArtifact extAOptionalDeploymentDep = TsArtifact.jar("ext-a-optional-deployment-dep");
+
+        final TsQuarkusExt extA = new TsQuarkusExt("ext-a");
+        addToExpectedLib(extA.getRuntime());
+        extA.getRuntime()
+                .addDependency(extADep)
+                .addDependency(new TsDependency(extAOptionalDep, true));
+        extA.getDeployment()
+                .addDependency(new TsDependency(extAOptionalDeploymentDep, true));
+
+        final TsArtifact extBOptionalDep = TsArtifact.jar("ext-b-optional-dep");
+        final TsArtifact extBDeploymentDep = TsArtifact.jar("ext-b-deployment-dep");
+        install(extBDeploymentDep);
+        final TsArtifact extBOptionalDeploymentDep = TsArtifact.jar("ext-b-optional-deployment-dep");
+        install(extBOptionalDeploymentDep);
+
+        final TsQuarkusExt extB = new TsQuarkusExt("ext-b");
+        extB.getRuntime().addDependency(new TsDependency(extBOptionalDep, true));
+        addToExpectedLib(extB.getRuntime());
+        extB.getDeployment().addDependency(new TsDependency(extBOptionalDeploymentDep, true));
+        extB.getDeployment().addDependency(new TsDependency(extBDeploymentDep, false));
+        install(extB);
+
+        final TsArtifact appOptionalDep = TsArtifact.jar("app-optional-dep")
+                .addDependency(extB.getRuntime());
+        addToExpectedLib(appOptionalDep);
+
+        final TsQuarkusExt extC = new TsQuarkusExt("ext-c");
+        install(extC);
+
+        final TsQuarkusExt extD = new TsQuarkusExt("ext-d");
+        extD.getRuntime().addDependency(new TsDependency(extC.getRuntime(), true));
+        extD.getDeployment().addDependency(new TsDependency(extC.getDeployment(), true));
+        install(extD);
+        addToExpectedLib(extD.getRuntime());
+
+        return TsArtifact.jar("app")
+                .addDependency(extA, true)
+                .addDependency(new TsDependency(appOptionalDep, true))
+                .addDependency(extD.getRuntime());
+    }
+
+    @Override
+    protected void assertAppModel(AppModel appModel) throws Exception {
+        final Set<AppDependency> expected = new HashSet<>();
+        expected.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a-deployment", "1"), "compile", true));
+        expected.add(
+                new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-b-deployment-dep", "1"), "compile", true));
+        expected.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-b-deployment", "1"), "compile", true));
+        expected.add(
+                new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-d-deployment", "1"), "compile", false));
+        assertEquals(expected, new HashSet<>(appModel.getDeploymentDependencies()));
+    }
+}

--- a/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/ProvidedExtensionDepsTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/runnerjar/ProvidedExtensionDepsTest.java
@@ -1,0 +1,48 @@
+package io.quarkus.deployment.runnerjar;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import io.quarkus.bootstrap.model.AppArtifact;
+import io.quarkus.bootstrap.model.AppDependency;
+import io.quarkus.bootstrap.model.AppModel;
+import io.quarkus.bootstrap.resolver.TsArtifact;
+import io.quarkus.bootstrap.resolver.TsDependency;
+import io.quarkus.bootstrap.resolver.TsQuarkusExt;
+
+public class ProvidedExtensionDepsTest extends ExecutableOutputOutcomeTestBase {
+
+    @Override
+    protected TsArtifact modelApp() {
+
+        final TsArtifact extADep = TsArtifact.jar("ext-a-dep");
+        addToExpectedLib(extADep);
+
+        final TsArtifact extAProvidedDep = TsArtifact.jar("ext-a-provided-dep");
+
+        final TsArtifact extADeploymentDep = TsArtifact.jar("ext-a-deployment-dep");
+        final TsArtifact extAOptionalDeploymentDep = TsArtifact.jar("ext-a-provided-deployment-dep");
+
+        final TsQuarkusExt extA = new TsQuarkusExt("ext-a");
+        addToExpectedLib(extA.getRuntime());
+        extA.getRuntime()
+                .addDependency(new TsDependency(extADep))
+                .addDependency(new TsDependency(extAProvidedDep, "provided"));
+        extA.getDeployment()
+                .addDependency(new TsDependency(extADeploymentDep))
+                .addDependency(new TsDependency(extAOptionalDeploymentDep, "provided"));
+
+        return TsArtifact.jar("app")
+                .addDependency(extA);
+    }
+
+    @Override
+    protected void assertAppModel(AppModel appModel) throws Exception {
+        final Set<AppDependency> expected = new HashSet<>();
+        expected.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a-deployment", "1"), "compile"));
+        expected.add(new AppDependency(new AppArtifact("io.quarkus.bootstrap.test", "ext-a-deployment-dep", "1"), "compile"));
+        assertEquals(expected, new HashSet<>(appModel.getDeploymentDependencies()));
+    }
+}

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -418,6 +418,9 @@ public class DevMojo extends AbstractMojo {
 
     private void triggerPrepare() throws MojoExecutionException {
         Plugin quarkusPlugin = project.getPlugin(QUARKUS_PLUGIN_GROUPID + ":" + QUARKUS_PLUGIN_ARTIFACTID);
+        if (quarkusPlugin == null) {
+            return;
+        }
         executeMojo(
                 quarkusPlugin,
                 goal(QUARKUS_PREPARE_GOAL),

--- a/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/CollectDependenciesBase.java
+++ b/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/CollectDependenciesBase.java
@@ -60,17 +60,21 @@ public abstract class CollectDependenciesBase extends ResolverSetupCleanup {
     }
 
     protected TsArtifact install(TsArtifact dep, String collectedInScope) {
-        return install(dep, null, collectedInScope);
+        return install(dep, null, collectedInScope, false);
+    }
+
+    protected TsArtifact install(TsArtifact dep, String collectedInScope, boolean optional) {
+        return install(dep, null, collectedInScope, optional);
     }
 
     protected TsArtifact install(TsArtifact dep, Path p, boolean collected) {
-        return install(dep, p, collected ? "compile" : null);
+        return install(dep, p, collected ? "compile" : null, false);
     }
 
-    protected TsArtifact install(TsArtifact dep, Path p, String collectedInScope) {
+    protected TsArtifact install(TsArtifact dep, Path p, String collectedInScope, boolean optional) {
         install(dep, p);
         if (collectedInScope != null) {
-            addCollectedDep(dep, collectedInScope, false);
+            addCollectedDep(dep, collectedInScope, optional);
         }
         return dep;
     }

--- a/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/TsArtifact.java
+++ b/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/TsArtifact.java
@@ -113,11 +113,15 @@ public class TsArtifact {
     }
 
     public TsArtifact addDependency(TsQuarkusExt dep) {
+        return addDependency(dep, false);
+    }
+
+    public TsArtifact addDependency(TsQuarkusExt dep, boolean optional) {
         if (extDeps.isEmpty()) {
             extDeps = new ArrayList<>(1);
         }
         extDeps.add(dep);
-        return addDependency(new TsDependency(dep.getRuntime()));
+        return addDependency(new TsDependency(dep.getRuntime(), optional));
     }
 
     public TsArtifact addDependency(TsDependency dep) {

--- a/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/test/OnlyDirectOptionalDepsAreCollectedTestCase.java
+++ b/independent-projects/bootstrap/core/src/test/java/io/quarkus/bootstrap/resolver/test/OnlyDirectOptionalDepsAreCollectedTestCase.java
@@ -15,10 +15,14 @@ public class OnlyDirectOptionalDepsAreCollectedTestCase extends CollectDependenc
         installAsDep(new TsDependency(new TsArtifact("required-dep-a"), "compile"), true);
         installAsDep(new TsDependency(new TsArtifact("required-dep-b"), "runtime"), true);
 
+        final TsArtifact nonOptionalCommon = new TsArtifact("non-optional-common");
+        install(nonOptionalCommon, "compile", true);
+
         installAsDep(
                 new TsDependency(
                         new TsArtifact("optional-dep")
                                 .addDependency(new TsDependency(new TsArtifact("common", "1"), true))
+                                .addDependency(new TsDependency(nonOptionalCommon, false))
                                 .addDependency(new TsDependency(new TsArtifact("other", "1"), true)),
                         true),
                 true);

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java
@@ -221,9 +221,10 @@ public class BootstrapAppModelResolver implements AppModelResolver {
         final List<RemoteRepository> repos = mvn.aggregateRepositories(managedRepos,
                 mvn.newResolutionRepositories(appArtifactDescr.getRepositories()));
 
-        final DeploymentInjectingDependencyVisitor deploymentInjector = new DeploymentInjectingDependencyVisitor(mvn,
-                managedDeps, repos, appBuilder);
+        final DeploymentInjectingDependencyVisitor deploymentInjector;
         try {
+            deploymentInjector = new DeploymentInjectingDependencyVisitor(mvn,
+                    managedDeps, repos, appBuilder);
             deploymentInjector.injectDeploymentDependencies(resolvedDeps);
         } catch (BootstrapDependencyProcessingException e) {
             throw new AppModelResolverException(

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/BuildDependencyGraphVisitor.java
@@ -136,10 +136,7 @@ public class BuildDependencyGraphVisitor {
         if (deploymentNode != null) {
             if (runtimeNode == null && !appDeps.contains(new AppArtifactKey(artifact.getGroupId(),
                     artifact.getArtifactId(), artifact.getClassifier(), artifact.getExtension()))) {
-                //we never want optional deps on the deployment CP
-                if (!node.getDependency().isOptional()) {
-                    deploymentDepNodes.add(node);
-                }
+                deploymentDepNodes.add(node);
             } else if (runtimeNode == node) {
                 runtimeNode = null;
                 runtimeArtifact = null;

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentDependencySelector.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentDependencySelector.java
@@ -1,0 +1,28 @@
+package io.quarkus.bootstrap.resolver.maven;
+
+import org.eclipse.aether.collection.DependencyCollectionContext;
+import org.eclipse.aether.collection.DependencySelector;
+import org.eclipse.aether.graph.Dependency;
+
+class DeploymentDependencySelector implements DependencySelector {
+
+    private final DependencySelector delegate;
+
+    public DeploymentDependencySelector(DependencySelector delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public boolean selectDependency(Dependency dependency) {
+        if (dependency.isOptional()) {
+            return false;
+        }
+        return delegate.selectDependency(dependency);
+    }
+
+    @Override
+    public DependencySelector deriveChildSelector(DependencyCollectionContext context) {
+        final DependencySelector newDelegate = delegate.deriveChildSelector(context);
+        return newDelegate == null ? null : new DeploymentDependencySelector(newDelegate);
+    }
+}

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java
@@ -15,6 +15,7 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Properties;
+import org.eclipse.aether.DefaultRepositorySystemSession;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.DefaultArtifact;
 import org.eclipse.aether.graph.Dependency;
@@ -48,8 +49,24 @@ public class DeploymentInjectingDependencyVisitor {
     private final AppModel.Builder appBuilder;
 
     public DeploymentInjectingDependencyVisitor(MavenArtifactResolver resolver, List<Dependency> managedDeps,
-            List<RemoteRepository> mainRepos, AppModel.Builder appBuilder) {
-        this.resolver = resolver;
+            List<RemoteRepository> mainRepos, AppModel.Builder appBuilder) throws BootstrapDependencyProcessingException {
+        // we need to be able to take into account whether the deployment dependencies are on an optional dependency branch
+        // for that we are going to use a custom dependency selector and re-initialize the resolver to use it
+        final DefaultRepositorySystemSession session = new DefaultRepositorySystemSession(resolver.getSession());
+        final DeploymentDependencySelector depSelector = new DeploymentDependencySelector(session.getDependencySelector());
+        session.setDependencySelector(depSelector);
+        try {
+            this.resolver = new MavenArtifactResolver(new BootstrapMavenContext(BootstrapMavenContext.config()
+                    .setRepositorySystem(resolver.getSystem())
+                    .setRepositorySystemSession(session)
+                    .setRemoteRepositories(resolver.getRepositories())
+                    .setRemoteRepositoryManager(resolver.getRemoteRepositoryManager())
+                    .setWorkspaceDiscovery(false)));
+        } catch (BootstrapMavenException e) {
+            throw new BootstrapDependencyProcessingException("Failed to initialize deployment dependencies resolver",
+                    e);
+        }
+
         this.managedDeps = managedDeps.isEmpty() ? new ArrayList<>() : managedDeps;
         this.mainRepos = mainRepos;
         this.appBuilder = appBuilder;

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenArtifactResolver.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/MavenArtifactResolver.java
@@ -126,6 +126,10 @@ public class MavenArtifactResolver {
         this.remoteRepoManager = mvnSettings.getRemoteRepositoryManager();
     }
 
+    public RemoteRepositoryManager getRemoteRepositoryManager() {
+        return remoteRepoManager;
+    }
+
     public MavenLocalRepositoryManager getLocalRepositoryManager() {
         return localRepoManager;
     }

